src/FshCompletionProvider.ts

```diff
@@ -131,6 +138,199 @@ export class FshCompletionProvider implements CompletionItemProvider {
+          } else if (infoToUse.type === 'Logical') {
+            entityToCheck = 'Base';
+          }
+        }
+      }
+    }
+    // is the entity present in the dependencies' entities?
+    // we may have to search twice, if we find a Profile
+    let originalType: string;
+    for (const [, entities] of this.fhirEntities) {
+      if (entities.profiles.has(entityToCheck)) {
+        originalType = entities.profiles.get(entityToCheck).baseDefinition;
+        break;
+      }
+      if (entities.extensions.has(entityToCheck)) {
+        return entities.extensions.get(entityToCheck).elements;
+      }
+      if (entities.resources.has(entityToCheck)) {
+        return entities.resources.get(entityToCheck).elements;
+      }
+      if (entities.logicals.has(entityToCheck)) {
+        return entities.logicals.get(entityToCheck).elements;
+      }
+    }
+    if (originalType != null) {
+      // search again, but don't include Profiles
+      for (const [, entities] of this.fhirEntities) {
+        if (entities.extensions.has(originalType)) {
+          return entities.extensions.get(originalType).elements;
+        }
+        if (entities.resources.has(originalType)) {
+          return entities.resources.get(originalType).elements;
+        }
+        if (entities.logicals.has(originalType)) {
+          return entities.logicals.get(originalType).elements;
+        }
+      }
+    }
+    return null;
+  }
+
+  public getPathItems(existingPath: string[], baseElements: ElementInfo[]): CompletionItem[] {
+    let targetElement: ElementInfo;
+    let targetLevel = baseElements;
+    for (const pathPart of existingPath) {
+      // if we are at a leaf element, try to expand based on types
+      if (targetLevel.length === 0 && targetElement.types?.length > 0) {
+        targetElement.children = this.generateItemsFromTypes(targetElement.types);
+        targetLevel = targetElement.children;
+      }
+      targetElement = targetLevel.find(element => element.path === pathPart);
+      targetLevel = targetElement?.children;
+      if (targetLevel == null) {
+        return [];
+      }
+    }
+    if (targetLevel) {
+      // if we are at a leaf element, try to expand based on types
+      if (targetLevel.length === 0 && targetElement.types?.length > 0) {
+        targetElement.children = this.generateItemsFromTypes(targetElement.types);
+        targetLevel = targetElement.children;
+      }
+      return targetLevel.map(element => new CompletionItem(element.path));
+    } else {
+      return [];
+    }
+  }
+
+  public generateItemsFromTypes(types: string[]): ElementInfo[] {
+    if (types.length === 0) {
+      return [];
+    }
+    const availableElements = types.map((type: string) => this.getBaseDefinitionElements(type));
+    if (availableElements.length === 1) {
+      return availableElements[0];
+    } else {
+      // on an element with multiple types, only include elements that can be used regardless of type
```

Comment:
Again, probably just me, but it took me some thinking to understand this.  It totally makes sense now that I understand it, but I think I would have got there quicker if it said:
```js
// on an element with multiple types, only return the elements that all types have in common
```
It may just be me.  I think I am having a difficult brain day.  If you disagree with this suggestion, feel free to disregard.